File: src/gop/Adapters/FreeProblemSet/Packer.cs
using gop.Helpers;
using gop.Problems;
using Markdig;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Xml;
using static gop.Helpers.TextIO;
using static gop.Helpers.ConsoleUI;
using PPipeline = gop.Adapters.Pipeline<gop.Problems.ProblemPath, string>;

namespace gop.Adapters.FreeProblemSet
{
    public static class Packer
    {
        public static XmlDocument GenerateFPS(ProblemPath problem, ProblemProfile config)
        {
            var xml = new XmlDocument();
            xml.AppendChild(xml.CreateXmlDeclaration("1.0", "UTF-8", null));
            var root = xml.CreateElement("fps");
            root.SetAttribute("url", "https://github.com/zhblue/freeproblemset/");
            root.SetAttribute("version", "1.2");

            {
                var gen = xml.CreateElement("generator");
                gen.SetAttribute("name", "HUSTOJ");
                gen.SetAttribute("url", "https://github.com/zhblue/hustoj/");
                root.AppendChild(gen);
            }

            {
                var item = xml.CreateElement("item");
                {
                    var sub = xml.CreateElement("title");
                    sub.AppendChild(xml.CreateCDataSection(config.Name));
                    item.AppendChild(sub);
                }
                {
                    var sub = xml.CreateElement("time_limit");
                    sub.SetAttribute("unit", "s");
                    sub.AppendChild(xml.CreateCDataSection(config.TimeLimit.ToString()));
                    item.AppendChild(sub);
                }
                {
                    var sub = xml.CreateElement("memory_limit");
                    sub.SetAttribute("unit", "mb");
                    sub.AppendChild(xml.CreateCDataSection(config.MemoryLimit.ToString()));
                    item.AppendChild(sub);
                }
                {
                    var sub = xml.CreateElement("description");
                    sub.AppendChild(xml.CreateCDataSection(Markdown.ToHtml(ReadAll(problem.Description))));
                    item.AppendChild(sub);
                }
                {
                    var sub = xml.CreateElement("input");
                    sub.AppendChild(xml.CreateCDataSection(Markdown.ToHtml(ReadAll(problem.Input))));
                    item.AppendChild(sub);
                }
                {
                    var sub = xml.CreateElement("output");
                    sub.AppendChild(xml.CreateCDataSection(Markdown.ToHtml(ReadAll(problem.Output))));
                    item.AppendChild(sub);
                }
                {
                    foreach (var u in problem.GetSamples())
                    {
                        var subin = xml.CreateElement("sample_input");
                        using (StreamReader sr = File.OpenText(u.InputFile))
                            subin.AppendChild(xml.CreateCDataSection(ConvertToLF(sr)));
                        var subout = xml.CreateElement("sample_output");
                        using (StreamReader sr = File.OpenText(u.OutputFile))
                            subout.AppendChild(xml.CreateCDataSection(ConvertToLF(sr)));
                        item.AppendChild(subin);
                        item.AppendChild(subout);
                    }
                }
                {
                    foreach (var u in problem.GetTests())
                    {
                        var subin = xml.CreateElement("test_input");
                        using (StreamReader sr = File.OpenText(u.InputFile))
                            subin.AppendChild(xml.CreateCDataSection(ConvertToLF(sr)));
                        var subout = xml.CreateElement("test_output");
                        using (StreamReader sr = File.OpenText(u.OutputFile))
                            subout.AppendChild(xml.CreateCDataSection(ConvertToLF(sr)));
                        item.AppendChild(subin);
                        item.AppendChild(subout);
                    }
                }
                {
                    var sub = xml.CreateElement("hint");
                    sub.AppendChild(xml.CreateCDataSection(Markdown.ToHtml(ReadAll(problem.Hint))));
                    item.AppendChild(sub);
                }
                {
                    var sub = xml.CreateElement("source");
                    sub.AppendChild(xml.CreateCDataSection(ReadAll(problem.Source)));
                    item.AppendChild(sub);
                }
                {
                    var sub = xml.CreateElement("solution");
                    sub.SetAttribute("language", "C++");
                    sub.AppendChild(xml.CreateCDataSection(ReadAll(problem.StandardProgram)));
                    item.AppendChild(sub);
                }
                root.AppendChild(item);
            }
            xml.AppendChild(root);
            return xml;
        }

        public const string LogCategory = "FreeProblemSet.Packer";
        public const string Token = "FreeProblemSet.Packer";

        public static PPipeline UseInitial(this PPipeline pipeline, PackageProfile package)
        {
            pipeline.SetToken(Token);
            pipeline.Result = null;
            package.Platform = "freeproblemset";
            return pipeline.UseCreate(package);
        }

        public static PPipeline UseDefault(this PPipeline pipeline, PackageProfile package)
        {
            return pipeline.UseInitial(package).UseFPS();
        }

        public static PPipeline UseCreate(this PPipeline pipeline, PackageProfile package)
        {
            const string LogCategory = Packer.LogCategory + ".Create";

            return pipeline.Use((pipe, problem) =>
            {
                WriteInfo(new OutputText("Packing the problem...", true));
                var logger = pipe.Logger;
                logger?.Info("Starting", LogCategory);
                var config = problem.GetProfile();
                pipe.Container.Set(config);
                pipe.Container.Set(package);
                Directory.CreateDirectory(problem.Target);

                // if (Directory.Exists(problem.Temp)) Directory.Delete(problem.Temp, true);
                // Directory.CreateDirectory(problem.Temp);

                string _file = $"{package.Platform}-{config.Author}-{config.Name}" + (package.Checked ? "" : "-unchecked") + ".xml";
                string filename = Path.Join(problem.Target, _file);
                if (File.Exists(filename)) File.Delete(filename);

                pipe.Container.Set(filename);

                logger?.Info("Ended", LogCategory);

                return problem;
            });
        }

        public static PPipeline UseFPS(this PPipeline pipeline)
        {
            const string LogCategory = Packer.LogCategory + ".FPS";

            return pipeline.Use((pipe, problem) =>
            {
                Write(new OutputText("  Create FPS data...", true));
                var logger = pipe.Logger;
                logger?.Info("Starting", LogCategory);

                var outputFile = pipe.Container.Get<string>();
                var config = pipe.Container.Get<ProblemProfile>();

                var xml = GenerateFPS(problem, config);

                using (StreamWriter sw = new StreamWriter(outputFile, false, UTF8WithoutBOM))
                    xml.Save(sw);

                logger?.Info("Ended", LogCategory);

                return problem;
            });
        }

        public static PPipeline UseSave(this PPipeline pipeline)
        {
            const string LogCategory = Packer.LogCategory + ".Save";

            return pipeline.Use((pipe, problem) =>
            {
                Write(new OutputText("  Saving...", true));
                var logger = pipe.Logger;
                logger?.Info("Starting", LogCategory);

                pipe.Result = pipe.Container.Get<string>();

                logger?.Info("Ended", LogCategory);

                return problem;
            });
        }        
    }
}

File: src/generator_oj_problem/adapters/fps/packer.py
from datetime import datetime
from glob import glob
import os
from pathlib import Path
from typing import Iterable
from generator_oj_problem.models import Issue, Problem, Severity, TestCase
from generator_oj_problem.pipelines import Reader, Packer
from generator_oj_problem import __version__
from yaml import safe_load
from xml.etree.ElementTree import Element, ElementTree, SubElement
from xml.dom.minidom import Document
import mistune


class Fps(Packer):
    def markdown(self, content: str):
        md = mistune.create_markdown(
            plugins=["url", "task_lists", "def_list", "abbr"])
        return md(content)

    def pack(self, reader: Reader, dist: Path) -> "Iterable[Issue]":
        problem = Problem()
        print("Load problem...")
        yield from reader.load(problem)

        doc = Document()

        root = doc.createElement("fps")
        doc.appendChild(root)
        root.setAttribute("url", "https://github.com/zhblue/freeproblemset/")
        root.setAttribute("version", "1.2")

        gen = doc.createElement("generator")
        root.appendChild(gen)
        gen.setAttribute("name", "HUSTOJ")
        gen.setAttribute("url", "https://github.com/zhblue/hustoj/")

        item = doc.createElement("item")
        root.appendChild(item)

        print("Pack description...")

        sub = doc.createElement("title")
        item.appendChild(sub)
        sub.appendChild(doc.createCDATASection(problem.name))

        sub = doc.createElement("time_limit")
        item.appendChild(sub)
        sub.setAttribute("unit", "s")
        sub.appendChild(doc.createCDATASection(str(int(problem.time))))

        sub = doc.createElement("memory_limit")
        item.appendChild(sub)
        sub.setAttribute("unit", "mb")
        sub.appendChild(doc.createCDATASection(str(int(problem.memory))))

        sub = doc.createElement("description")
        item.appendChild(sub)
        sub.appendChild(doc.createCDATASection(
            self.markdown(problem.description)))

        sub = doc.createElement("input")
        item.appendChild(sub)
        sub.appendChild(doc.createCDATASection(self.markdown(problem.input)))

        sub = doc.createElement("output")
        item.appendChild(sub)
        sub.appendChild(doc.createCDATASection(self.markdown(problem.output)))

        print("Pack sample data...")

        for case in reader.samples():
            print(f"  Pack sample {case.name}...")
            subin = doc.createElement("sample_input")
            subin.appendChild(doc.createCDATASection(
                "\n".join(case.input.splitlines()) + "\n"))

            subout = doc.createElement("sample_output")
            subout.appendChild(doc.createCDATASection(
                "\n".join(case.output.splitlines()) + "\n"))

            item.appendChild(subin)
            item.appendChild(subout)

        print("Pack test data...")

        for case in reader.tests():
            print(f"  Pack test {case.name}...")
            subin = doc.createElement("test_input")
            subin.appendChild(doc.createCDATASection(
                "\n".join(case.input.splitlines()) + "\n"))

            subout = doc.createElement("test_output")
            subout.appendChild(doc.createCDATASection(
                "\n".join(case.output.splitlines()) + "\n"))

            item.appendChild(subin)
            item.appendChild(subout)

        print("Pack extra...")

        hint = f"""{problem.hint}

*Generated at {datetime.now()} by [generator-oj-problem](https://github.com/StardustDL/generator-oj-problem) v{__version__}.*
"""

        if not hint.isspace():
            sub = doc.createElement("hint")
            item.appendChild(sub)
            sub.appendChild(doc.createCDATASection(
                self.markdown(hint)))

        if not problem.author.isspace():
            sub = doc.createElement("source")
            item.appendChild(sub)
            sub.appendChild(doc.createCDATASection(problem.author))

        if not problem.solution.isspace():
            sub = doc.createElement("solution")
            item.appendChild(sub)
            sub.setAttribute("language", problem.solutionLanguage)
            sub.appendChild(doc.createCDATASection(problem.solution))

        print("Save dist...")
        try:
            fp = dist / "fps.xml"
            with open(fp, "w", encoding="utf-8") as f:
                doc.writexml(f, encoding="utf-8")
            yield Issue(f"Saved to {fp}.")
        except Exception as ex:
            yield Issue(f"Failed to save: {ex}", Severity.Error)
